### PR TITLE
Automatic support for META-INF/services

### PR DESCRIPTION
**Pull request by jtulach**
In my projects I use java.util.ServiceLoader quite a lot. The common usage pattern looks like this:
```
for (SomeInterface impl : ServiceLoader.load(SomeInterface.class)) {
  // do something with impl
}
```
However this pattern doesn't work well with RoboVM's class closure computation. There is no direct dependency on the implementation classes, so RoboVM does not include them in its closure automatically. But it could, the behaviour of ServiceLoader is quite strict and one could read the same META-INF/services/pkg.SomeInterface files and include them in the class closure as well.

I am attaching my changes which allow me to remove few of my 

```
 <forceLinkClasses>
  <pattern>org.netbeans.html.ko4j.KO4J</pattern>
  <pattern>org.netbeans.html.sound.impl.BrowserAudioEnv</pattern>
</forceLinkClasses>
```
and thus simplify my build scripts. Would you mind accepting them?

### CODE REVIEW

File: compiler/src/main/java/org/robovm/compiler/AppCompiler.java [hunk 1]
```diff
@@ -241,10 +244,33 @@ private boolean compile(Executor executor, ClassCompilerListener listener,
...
                 }
             }
+            addMetaInfImplementations(config.getClazzes(), clazz, compiled, compileQueue);
         }
         return result;
     }
 
+    static void addMetaInfImplementations(Clazzes clazzes, Clazz clazz, Set<Clazz> compiled, Set<Clazz> compileQueue) throws IOException {
+        String metaInfName = "META-INF/services/" + clazz.getClassName();
+        for (InputStream is : clazzes.loadResources(metaInfName)) {
```
Comment (badlogic) [inline compiler/src/main/java/org/robovm/compiler/AppCompiler.java]: If parsing one of the streams goes wrong, the remaining streams aren't closed. May get cleaned up in finalizes eventually, but given enough heap that might never happen and we could run out of file handles.

**Comment by badlogic** [comment]
Looks good. We do already have a mechanism for libraries to declare force linked classes and other RoboVM specific configuration. Putting a robovm.xml in the META-INF path of a library does the trick. See the robovm-ios-bindings. The parent build.gradle does the packaging: https://github.com/BlueRiverInteractive/robovm-ios-bindings/blob/master/build.gradle

**Comment by jtulach** [comment]
I was assuming that once an IOException occurs, the compilation ends anyway, but you are right, the code should be more robust. It is not easy to shield against all errors properly - e.g. to mimic try-with-resources - the best way I could think of was recursive nesting of try-with-resources, but then the code would become a bit unreadable. As such I ended up with check against IOException only. I can improve it against other types of error, if requested.

**Comment by jtulach** [comment]
Unless I am mistaken there is a difference in "laziness" between robovm.xml approach and the proposed processing of META-INF/services. In the services approach one needs to use the interface first, only then  the implementation is included. With forceLinkClass, one includes the implementation (and the interface, and its closure) regardless whether it is needed or not.

**Comment by badlogic** [comment]
Ah, that makes sense. Doesn't solve the issue when the class is loaded via reflection at runtime. The PR looks good for to me otherwise. @ntherning will have to decide on whether this should be merged.

File: compiler/src/main/java/org/robovm/compiler/AppCompiler.java [hunk 2]
```diff
@@ -241,10 +244,39 @@ private boolean compile(Executor executor, ClassCompilerListener listener,
...
+            addMetaInfImplementations(config.getClazzes(), clazz, compiled, compileQueue);
         }
         return result;
     }
 
+    static void addMetaInfImplementations(Clazzes clazzes, Clazz clazz, Set<Clazz> compiled, Set<Clazz> compileQueue) throws IOException {
+        String metaInfName = "META-INF/services/" + clazz.getClassName();
+        IOException throwLater = null;
+        for (InputStream is : clazzes.loadResources(metaInfName)) {
+            try (BufferedReader r = new BufferedReader(new InputStreamReader(is))) {
```
Comment (ntherning) [inline compiler/src/main/java/org/robovm/compiler/AppCompiler.java]: Perhaps use an explicit charset name here? UTF-8? Unlikely to become a problem unless the user has some weird default charset like UTF-16.

File: compiler/src/test/java/org/robovm/compiler/AppCompilerTest.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright (C) 2015 Trillian Mobile AB
```
Comment (ntherning) [inline compiler/src/test/java/org/robovm/compiler/AppCompilerTest.java]: We recently changed company name to "RoboVM AB"

**Comment by ntherning** [comment]
This is very smart! I will happily merge this, just see the very minor comments I had. And also, we will need a signed CLA before we can merge it. Please see http://robovm.com/cla. Thanks!

**Comment by jtulach** [comment]
I've submitted the CLA. I can fix the header issue when I get back home to my Mac - later tomorrow - or feel free to change it when merging.

**Comment by ntherning** [comment]
Thanks!

**Pull request merged**